### ISSUE TITLE
Support for unorm and snorm type qualification

### ISSUE THREAD

**Issue by jsmall-nvidia** [shader-slang/slang#935]
When defining uav resource, sometimes it is necessary to qualify the underlying type such that correct results are produced. HLSL has support for two qualifications 'unorm' and 'snorm'. For example

```
RWTexture2D<unorm float4> geometryNormal;
```

The issue is covered here https://docs.microsoft.com/en-us/windows/desktop/direct3d12/typed-unordered-access-view-loads#using_unorm_and_snorm_typed_uav_loads_from_hlsl

At the end of the document it states... 

> Using UNORM and SNORM typed UAV loads from HLSL
>
> When using typed UAV loads to read from a UNORM or SNORM resource, you must properly declare the element type of the HLSL object to be unorm or snorm. It is specified as undefined behavior to mismatch the element type declared in HLSL with the underlying resource data type. For example, if you are using typed UAV loads on a buffer resource with R8_UNORM data, then you must declare the element type as unorm float:

Currently slang does not support these qualifications and so doing so will produce an error. 

snorm means the value will be in the range -1 to 1, and unorm means the range 0 to 1. In HLSL you can define variables with the qualification - but it does not appear to have any checking or clamping. For example...

```
unorm float a = 2.0f;
// Seems equivalent to (if compiled on fxc or dxc) 
float a = 2.0f;
```

The following link just states

https://docs.microsoft.com/en-us/windows/desktop/direct3dhlsl/dx-graphics-hlsl-scalar

```
    snorm float - IEEE 32-bit signed-normalized float in range -1 to 1 inclusive.
    unorm float - IEEE 32-bit unsigned-normalized float in range 0 to 1 inclusive.
```

For GL the unorm/snorm types are available (snorm through an extension and implicitly for 8 bit types). For Vulkan VkFormat supports unorm and snorm types.  

https://www.khronos.org/opengl/wiki/GLAPI/glTexImage2D
https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkFormat.html

It does not appear extra qualification is needed in glsl in the shader. For example the snorm extension documentation doesn't mention anything about glsl changes. 

https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_texture_snorm.txt

To implement in slang we may want to extend QualType such that it holds unorm or snorm. It should allow conversions as if the qualification is not specified (to match dxc/fxc).  On emit we need to honor the qualification on output. Also needed is an AST node case to represent a qualified type expression (e.g., a `QualTypeExpr : Expr`), and an IR type say 'IRQualType'.

**Comment by jsmall-nvidia** [comment]
Some follow up testing on this shader...

```
struct PSInput
{
	float4 color : COLOR;
};
Texture2D<snorm float4> gTexture;
RWTexture2D<snorm float> gOutTexture;
float4 PSMain(PSInput input) : SV_TARGET
{    
        gOutTexture[uint2(input.color.xy)] = input.color.x;
	return gTexture[uint2(input.color.xy)];
}
```

The test was in comparing with snorm replaced by unorm or nothing. I compared the binaries.

With DXIL PS_6 on DXC the output does not change.
With SPIR-V on Vulkan 1.0 the output does not change.

With FXC the output does change - and the unorm and snorm appear as 

```dcl_resource_texture2d (snorm,snorm,snorm,snorm) t0
dcl_uav_typed_texture2d (snorm,snorm,snorm,snorm) u1```

and 

```
ld_indexable(texture2d)(snorm,snorm,snorm,snorm) o0.xyzw, r0.xyzw, t0.xyzw
``` 

More experimentation on fxc shows that unorm and snorm are very limited in their uses. For example 

```
typedef float4 Float4Type;
Texture2D<unorm Float4Type> gTexture;
```

In a declaration it must be a built in type it seems. On DXC this isn't the case and this works (although makes no difference to the output).

**Comment by tfoleyNV** [comment]
One additional note on fxc behavior is that if I declare a function like:

```hlsl
float4 f(Texture2D<float4> t) { ... }
```

and then try to call it on a `Texture2D<unorm float4>`, I get an error:

```
error X3017: 'f': cannot implicitly convert from 'const Texture2D<unorm float4>' to 'Texture2D<float4>'
```

So the most consistent answer would seem to be that we'd treat `unorm T` and `snorm T` as distinct types from `T`, but with an implicit conversion defined both to and from `T`. This would guarantee that `Texture2D<unorm T>` is a distinct type from `Texture2D<T>`, while still allowing the result of `.Sample` on a `Texture2D<uniform T>` to convert to `T`.

That solution would involve (in the simplest case) adding something like a `UNormType` and `SNormType` (with a common base class and a `baseType` field for the underlying type)  to the type system (at the AST and IR levels) an then adding logic to allow them to be implicitly converted to their base type. We'd also need a `UNormTypeExpr` and `SNormTypeExpr` for the parser to produce when it sees `unorm ...` in a type expression, and then `check.cpp` would handle resolving the base type expression and constructing the `UNormType` or `SNormType` (and could also double-check that the base type makes sense for that qualification).

During emit we'd then mostly just treat `unorm T` as `T`, and the only place we'd check for it explicitly is when emitting the element type of a `Texture*<...>` for HLSL (I guess we could also use it in the inference of formats for VK...).

That would be a relatively direct way to support this feature if we decide it is important, although it might be less elegant than adding a more generalized system for representing type qualifiers.

**Comment by tangent-vector** [comment]
I believe this is implemented in #2095.
